pkg/cloud/aws/services/ec2/subnets.go

```diff
@@ -175,14 +175,56 @@ func (s *Service) describeVpcSubnets() (v1alpha1.Subnets, error) {
 		return nil, errors.Wrapf(err, "failed to describe subnets in vpc %q", s.scope.VPC().ID)
 	}
 
+	routeTables, err := s.describeVpcRouteTablesBySubnet()
+	if err != nil {
+		return nil, err
+	}
+
+	natGateways, err := s.describeNatGatewaysBySubnet()
+	if err != nil {
+		return nil, err
+	}
+	internetGateways, err := s.describeVpcInternetGateways()
+	if err != nil {
+		return nil, err
+	}
+
 	subnets := make([]*v1alpha1.SubnetSpec, 0, len(out.Subnets))
 	for _, ec2sn := range out.Subnets {
+		rt := routeTables[*ec2sn.SubnetId]
+		var hasPublicRoute bool
+		var routeTableId, natGatewayId *string
+		if rt != nil {
+			routeTableId = rt.RouteTableId
+			for _, route := range rt.Routes {
+				if route == nil {
+					continue
+				}
+				if route.GatewayId != nil {
+					for _, igw := range internetGateways {
+						if igw == nil {
+							continue
+						}
+						if *route.GatewayId == *igw.InternetGatewayId {
+							hasPublicRoute = true
```

Comment:
What's the reasoning behind checking if the route's internet gateway is in the `internetGateways` slice? Can we assume that IGW associated with a subnet's routing table inside a VPC can only point to a gateway attached to the VPC?

Reply:
From what I could tell, determining the "type" of a route's destination is a string parsing/matching problem: "does this route point to an internet gateway?" can be translated into either:

- "does it match an id that's known to be an internet gateway?"
- "does it start with `igw`?"

I went the former route because I prefer when possible to treat IDs as opaque, but I guess Hyrum's law here would suggest that amazon isn't going to change their structure any time soon. Do you have a preference on which way to go?

Reply:
String matching on `igw` sounds good to me, for now only internet gateways are supported anyways :)